asyncstatsd/client/basic.py

```diff
@@ -0,0 +1,81 @@
+import contextlib
+import random
+from abc import ABC, abstractmethod
+from datetime import timedelta
+from time import perf_counter as time_now
+
+from ..metrics.basic import (
+    CounterMetric,
+    GaugeMetric,
+    SetMetric,
+    StatsdMetric,
+    TimingMetric,
+)
+from .udp import UDPClient
+
+
+class StatsdClientBase(UDPClient):
+    def __init__(self, host, port, *, maxudpsize=512, prefix: str = ""):
+        super().__init__(host, port, maxudpsize=maxudpsize)
+        self._prefix = prefix
+
+    def send(self, data: StatsdMetric):
+        serialized = "".join([self._prefix, data.serialize()])  # type: ignore
+        if data.rate < 1:
+            if random.random() > data.rate:
+                return
```

Comment:
What is the purpose of this check? It appears to randomly drop some metrics if their rate is below 1

Reply:
yes, if rate is 0.5 for example we are sampling metric on the client side i.e. dropping 1 out of 2 metrics to achieve required sampling rate

Reply:
oh nice. So it relies on the random numbers being uniformly distributed, which I imagine they are, pretty clever